proposals/NNNN-System-Request.md

```diff
@@ -0,0 +1,94 @@
+
+## Motivation
+
+With the growing popularity of SDL, more components within the head unit would like to exchange their data in a component specific proprietary way with the owner applications. Most of the time, the size of the information is less than a kilo-Bytes. Currently mobile apps can send/receive binary data in SystemRequest and OnSystemRequest RPC messages using hybrid message, a file needs to be passed between SDL and HMI to exchange the data. File operations (create, read, write, delete) are slow. It is better to send data (especially a small amount of data) using message directly.
+
+
+## Proposed solution
+
+There is no change to the mobile API. This proposal adds a new optional parameter to the existing `SystemRequest` request/response and `OnSystemRequest` notification.
+
+ 
+HMI API changes
+
+```xml
+<function name="OnSystemRequest" messagetype="notification" >
+    ...
++   <param name="data" type="String" maxlength="65535" mandatory="false">
++       <description> 
++           Base64 encoded data sending from HMI to SDL. If it exists, "fileName" shall be ignored; 
++           SDL shall decode the string and treat the resulting data like it comes from a binary file.
++           SDL shall send decoded data to the mobile app in hybrid part of message
++        </description>
++   </param>
+</function>
+
+
+<function name="SystemRequest" messagetype="request">
+    ...
++   <param name="requestData" type="String" maxlength="65535" mandatory="false">
++       <description>
++         If it exists, "fileName" shall be ignored;
++         If the length of incoming binary data from hybrid message is less than configured bytes, instead of saving it to a binary file, SDL base64 encoded the data.  
++       </description>
++   </param>
+</function>
+
+
+<function name="SystemRequest" messagetype="response">
+    ...
++   <param name="resultData" type="String" maxlength="65535" mandatory="false">
++       <description> Base64 encoded result data from the system to SDL. SDL shall send decoded data to the mobile app in hybrid part of message. </description>
++   </param>
+</function>
+```
+
+
+Here are the changes needed by SDL core.
+- Add a new configurable parameter in smartdevicelink.ini
+
+```
+; The payload in SystemRequest/OnSystemRequest packet bigger than this value will be saved to a file when passing to HMI,
+; payload smaller than this value will be based64 encoded and stored as a string parameter to HMI 
+; if not specified, this value will default to 0
+MaximumBinaryPayloadSizeToString = 11520
+```
+
+- When core receives a SystemRequest request from mobile, if the incoming binary data is less than a certain size specified by smartdevicelink.ini, instead of saving the data to a temporary file, SDL shall encode the data using base64 format, and send the encoded string directly within a SystemRequest HMI message. If the size of data is large, SDL saves the data to a file and pass the file name to HMI as before.
```

Comment:
```suggestion
- When SDL Core receives a SystemRequest request from mobile, if the incoming binary data is less than a certain size specified by smartdevicelink.ini, instead of saving the data to a temporary file, SDL Core shall encode the data using base64 format, and send the encoded string directly within a SystemRequest HMI message. If the size of data is large, SDL Core saves the data to a file and passes the file name to HMI as before.
```